### PR TITLE
Add support for source generators in Razor compiler

### PR DESCRIPTION
**Pull request by captainsafia**
- Update the version of `Microsoft.CodeAnalysis.CSharp` that we consume otherwise we get some downstream assembly resolution issues in the version of `Microsoft.AspNetCore.Razor.Extensions` that we use in the SDK repo.
- Add IVT info for M.NET.Sdk.Razor.SourceGenerators assembly. I'm working around this for now by giving the source generators package an `rzc` assembly name over in the sdk repo.
- Read from source document in project item. I'm working around this by implementing a `Read` method that returns a text stream from the string of a `SourceText` object but that's not ideal.

Part of https://github.com/dotnet/aspnetcore/issues/26902

### CODE REVIEW

**Review by pranavkm — APPROVED** [review]
I think the new compiler introduces some additional nullability warnings that you would need to address. But otherwise this looks good.

FYI @NTaylorMullen

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorProjectItem.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorProjectItem.cs:133]: Hmm, this is concerning. The intent of RazorProjectItem is to be decoupled from a SourceDocument entirely. Why do you need this here? Typically we just re-create a project item.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs:196]: Why not just re-create a source document?

Comment (captainsafia) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs:196]: This and the change above are designed to work in conjunction with some work we've done to support source generators.

We have a new `SourceGeneratorProjectItem` ([ref](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/SourceGeneratorProjectItem.cs)) type that gets constructed from the list of files that Roslyn provides in a generator's compilation context ([ref](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/RazorSourceGenerationContext.cs#L105)).

These additional files are provided as `SourceText`s. We have a `SourceTextRazorSourceDocument` type ([ref](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/SourceTextRazorSourceDocument.cs) that we create from the `SourceText`.

Here we are updating the `ReadFrom` method to directly use the `SourceDocument` that we compute from the `SourceText` provided by Roslyn instead of reading in the file contents from a stream.

You can view a hacky workaround I used to get around this [here](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/SourceGeneratorProjectItem.cs#L43) by reading the string of the `SourceText` into a stream but this isn't ideal.

Hope this explanation was clear. Thoughts?

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorProjectItem.cs
Comment (pranavkm) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorProjectItem.cs:133]: One of the things that helped with perf was having a `RazorSourceLineCollection` that was backed by the Roslyn `SourceText` document (because that's what we're provided in a source generator). Here's what that looks like: https://github.com/dotnet/aspnetcore/pull/28807/files#diff-04e8fa11176d9e8cdfa5d0edeeb3ba7c1c6d20683201775277a3215a1fcb1a3f. 

If I'm not mistaken, there isn't an API that lets you affect how a RazorProjectItem gets turned in to a RazorSourceDocument. We could create one where the default implementation reads from the stream and produces a source document and the source generator configures it with a custom provider. This one felt easier to implement as we did not have to introduce yet another concept, but you have more context here, so open to suggestions.

**Comment by dougbu** [comment]
@captainsafia what in particular would you like me to review❔ If it's just the CODEOWNERS file, no objections from me

/fyi dotnet/install-scripts#136 is blocking this PR at the moment

File: src/Components/Components/src/RenderTree/RenderTreeEdit.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/RenderTreeEdit.cs:56]: Should we change the declaration of this field to be nullable instead?

It not only can be null, it usually is null.

Comment (captainsafia) [reply src/Components/Components/src/RenderTree/RenderTreeEdit.cs]: I tried this at first but ran into some annoying issues with the public APIs analyzer. After a night's sleep, turns out we have to make some updates to the PublicShipped.API.txt files under both the Component and Ignitor project since we compile include source files from the components project in the ignitor project.

**Comment by dougbu** [comment]
> Turns out we have to make some updates to the PublicShipped.API.txt files under both the Component and Ignitor project since we compile include source files from the components project in the ignitor project.

Changing the PublicAPI.**Shipped**.txt files outside an X.Y.**0** release is _never_ the right idea. @JunTaoLuo is scrambling to fix up the places where we've made this mistake. Please don't make it worse.

If necessary, change the PublicAPI.**Un**shipped.txt files to remove changed APIs. The syntax is to copy the line from PublicAPI.Shipped.txt (again, not changing that file) into PublicAPI.Unshipped.txt and add a `*Removed*` prefix.

**Comment by pranavkm** [comment]
@dougbu in this case it's changing the nullability annotation on a shipped API. We've made a number of those so far - https://github.com/aspnet/Announcements/issues/444.

File: src/Components/Ignitor/src/PublicAPI.Shipped.txt
Comment (pranavkm) [inline src/Components/Ignitor/src/PublicAPI.Shipped.txt:193]: Hmm, we could turn off the PublicAPI analyzer on this project. We're not shipping this package.

File: src/Components/Components/src/PublicAPI.Shipped.txt
Comment (pranavkm) [inline src/Components/Components/src/PublicAPI.Shipped.txt:313]: Could you add this to https://github.com/aspnet/Announcements/issues/444?

**Review by pranavkm — APPROVED** [review]

**Comment by dougbu** [comment]
> We've made a number of those so far - aspnet/Announcements#444.

I know. The problem isn't the nullabiltity changes but how they were handled in the API baseline files. PublicAPI.Shipped.txt files should change only immediately after X.Y.0 ships.

File: src/Components/Components/src/PublicAPI.Shipped.txt
Comment (captainsafia) [reply src/Components/Components/src/PublicAPI.Shipped.txt:313]: Done.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs
Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs]: Thanks for the clarification, definitely helps piece it all together a bit more.  It's been a while since I've dug through the code but why can't we have the project item read directly from the Roslyn source text?

Comment (captainsafia) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs]: We did create a ProjectItem implementation that reads directly from the Roslyn source text ([ref](https://github.com/dotnet/sdk/blob/safia/rsg/src/RazorSdk/SourceGenerators/SourceGeneratorProjectItem.cs#L10)). But currently, the project engine computes the source document ([ref](https://github.com/dotnet/aspnetcore/blob/70a880f30cd1537ff3777f51efe18469c8da68db/src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorProjectEngine.cs#L78)) from a project item by reading from a stream ([ref](https://github.com/dotnet/aspnetcore/blob/70a880f30cd1537ff3777f51efe18469c8da68db/src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs#L196-L201)).

We can try implementing an overload of `ReadFrom` that takes in a `SourceGeneratorProjectItem`

```csharp
public static RazorSourceDocument ReadFrom(SourceGeneratorProjectItem projectItem)
{
   return new SourceGeneratorSourceDocument(projectItem.AdditionalText); // or similiar
}
```

But that would add source generator-specific logic in the compiler.

An alternative is to create some sort of abstraction where in a custom `ProjectItem` can define how source documents should be read from it similar to @pranavkm's recommendation above.

Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs]: > An alternative is to create some sort of abstraction where in a custom `ProjectItem` can define how source documents should be read from it similar to @pranavkm's recommendation above.

Ya I think that'd be better. It'd be a more defined route of translating from one type to another.

Comment (captainsafia) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/RazorSourceDocument.cs]: Yep! Will create an issue to track this. We want to get the initial E2E with source generators working then refine things.

**Pull request merged**
